includes/Modules/Site_Verification.php

```diff
@@ -336,4 +385,103 @@ protected function setup_services( Google_Client $client ) {
 	}
+
+	/**
+	 * Handles receiving a verification token for a user by the authentication proxy.
+	 *
+	 * @since n.e.x.t
+	 */
+	private function handle_verification_token() {
+		$authentication = $this->authentication;
+		$auth_client    = $authentication->get_oauth_client();
+
+		if ( ! $auth_client->using_proxy()
+			|| empty( $_GET['googlesitekit_verification_token'] )
+			|| empty( $_GET['googlesitekit_verification_nonce'] )
+		) {
+			return;
+		}
+
+		if ( ! wp_verify_nonce( $_GET['googlesitekit_verification_nonce'], 'googlesitekit_verification' ) ) {
+			wp_die( esc_html__( 'Invalid nonce.', 'google-site-kit' ) );
+		}
+
+		$verification_type = isset( $_GET['googlesitekit_verification_token_type'] ) ? $_GET['googlesitekit_verification_token_type'] : self::VERIFICATION_TYPE_META;
+		switch ( $verification_type ) {
+			case self::VERIFICATION_TYPE_FILE:
+				$authentication->verification_file()->set( $_GET['googlesitekit_verification_token'] );
+				break;
+			case self::VERIFICATION_TYPE_META:
+				$authentication->verification_tag()->set( $_GET['googlesitekit_verification_token'] );
+		}
+
+		$code = isset( $_GET['googlesitekit_code'] ) ? $_GET['googlesitekit_code'] : '';
+
+		// We need to pass the 'missing_verification' error code here so that the URL includes a verification nonce.
+		wp_safe_redirect(
+			add_query_arg(
+				array(
+					'verify'              => 'true',
+					'verification_method' => $verification_type,
+				),
+				$auth_client->get_proxy_setup_url( $code, 'missing_verification' )
+			)
+		);
+		exit;
+	}
+
+	/**
+	 * Prints site verification meta in wp_head().
+	 *
+	 * @since n.e.x.t
+	 */
+	private function print_site_verification_meta() {
+		// Get verification meta tags for all users.
+		$verification_tags = $this->authentication->verification_tag()->get_all();
+		$allowed_html      = array(
+			'meta' => array(
+				'name'    => array(),
+				'content' => array(),
+			),
+		);
+
+		foreach ( $verification_tags as $verification_tag ) {
+			$verification_tag = html_entity_decode( $verification_tag );
+
+			if ( 0 !== strpos( $verification_tag, '<meta ' ) ) {
+				$verification_tag = '<meta name="google-site-verification" content="' . esc_attr( $verification_tag ) . '">';
+			}
+
+			echo wp_kses( $verification_tag, $allowed_html );
+		}
+	}
+
+	/**
+	 * Serves the verification file response.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $verification_file_name File name of verification.
+	 */
+	private function serve_verification_file( $verification_file_name ) {
+		global $wpdb;
+
+		$user_ids = ( new \WP_User_Query(
+			array(
+				'meta_key'   => $wpdb->get_blog_prefix() . Verification_File::OPTION,
+				'meta_value' => $verification_file_name,
+				'fields'     => 'id',
+				'number'     => 1,
+			)
+		) )->get_results();
+
+		$user_id = array_shift( $user_ids ) ?: 0;
+
+		if ( user_can( $user_id, Permissions::SETUP ) ) {
+			printf( 'google-site-verification: %s', esc_html( $verification_file_name ) );
```

Comment:
At this point, the current functionality is actually broken because it also prints the `.html` suffix here. As soon as we update the logic as mentioned above to only expect the token itself from the proxy without `.html`, we should be okay.